planner/core/rule_aggregation_elimination.go

```diff
@@ -155,3 +272,174 @@ func (a *aggregationEliminator) optimize(p LogicalPlan) (LogicalPlan, error) {
 	}
 	return p, nil
 }
+
+func collectNestedAggPattern(lp LogicalPlan, ptn *nestedAggPattern) {
+	if len(lp.Children()) != 1 {
+		return
+	}
+	child := lp.Children()[0]
+	switch v := child.(type) {
+	case *LogicalAggregation:
+		ptn.inner = v
+		return
+	case *LogicalProjection:
+		if ptn.proj != nil {
+			return
+		}
+		ptn.proj = v
+	case *LogicalSort:
+		// Ignore sort between aggregations.
+	default:
+		ptn.isTrivial = false
+	}
+	collectNestedAggPattern(child, ptn)
+}
+
+// tryToCombineAggFunc checks the types of inner/outer aggregate function and check if they can be combined as one based on their semantics.
+// for example, since max(max(PARTIAL)) can be combined as max(TOTAL), we can combine inner max() and outer max() as a final max()
+// `innerIsGroup` indicates if the inner aggregate function is aggregating group-by items(values are de-duplicated).
+func tryToCombineAggFunc(outer, inner *aggregation.AggFuncDesc, innerIsGroup bool, expr expression.Expression) *aggregation.AggFuncDesc {
+	combined := outer.Clone()
+	combined.Args[0] = expr
+
+	if innerIsGroup {
+		if (inner.Name == ast.AggFuncFirstRow || inner.Name == ast.AggFuncMax || inner.Name == ast.AggFuncMin) &&
+			(outer.Name == ast.AggFuncFirstRow || outer.Name == ast.AggFuncMax || outer.Name == ast.AggFuncMin) {
+			return combined
+		}
+		// TODO: add type cast to support more cases
+	}
+
+	switch {
+	case inner.Name == ast.AggFuncFirstRow && outer.Name == ast.AggFuncFirstRow:
+	case inner.Name == ast.AggFuncMax && outer.Name == ast.AggFuncMax:
+	case inner.Name == ast.AggFuncMin && outer.Name == ast.AggFuncMin:
+	case inner.Name == ast.AggFuncBitAnd && outer.Name == ast.AggFuncBitAnd:
+	case inner.Name == ast.AggFuncBitOr && outer.Name == ast.AggFuncBitOr:
+	case inner.Name == ast.AggFuncBitXor && outer.Name == ast.AggFuncBitXor:
+	case inner.Name == ast.AggFuncGroupConcat && outer.Name == ast.AggFuncGroupConcat:
+		if inner.HasDistinct {
```

Comment:
`first_row / min / max / BitAnd / BitOr / BitXor` is applicable even if `inner.HasDistinct` is true? Also, `group_concat` cannot be simply merged because it may contain `ORDER BY` and `SEPARATOR` keyword inside the `group_concat`?

Reply:
`first_row`, `min`, `max`, `bit_and`, `bit_or` accept `HasDistinct` because they're idempotent. But `bit_xor` is not, I've fixed it, thanks!